watcher/local/watcher.go

```diff
@@ -102,13 +114,156 @@ func (w *Watcher) startWatching(
 
 	w.registry.addUnsafe(ch)
 
+	tx := channel.Transaction{
+		State: signedState.State,
+		Sigs:  signedState.Sigs,
+	}
+	go handleStatesFromClient(tx, statesPubSub, ch.requestLatestTx, ch.latestTx)
+	go w.handleEventsFromChain(eventsFromChainSub, eventsToClientPubSub, ch)
+
 	return statesPubSub, eventsToClientPubSub, nil
 }
 
 func newCh(id channel.ID, parent *ch, params *channel.Params) *ch {
 	return &ch{
-		id:     id,
-		params: params,
-		parent: parent,
+		id:                id,
+		params:            params,
+		parent:            parent,
+		registeredVersion: 0,
+		requestLatestTx:   make(chan struct{}),
+		latestTx:          make(chan channel.Transaction),
+	}
+}
+
+func (ch *ch) retreiveLatestTx() channel.Transaction {
+	ch.requestLatestTx <- struct{}{}
+	return <-ch.latestTx
+}
+
+func handleStatesFromClient(currentTx channel.Transaction, statesSub statesSub, requestLatestTxn chan struct{},
+	latestTx chan channel.Transaction) {
+	var _tx channel.Transaction
+	var ok bool
+	for {
+		select {
+		case _tx, ok = <-statesSub.statesStream():
+			if !ok {
+				log.WithField("ID", currentTx.State.ID).Info("States sub closed by client. Shutting down handler")
+				return
+			}
+			currentTx = _tx
+			log.WithField("ID", currentTx.ID).Debugf("Received state from client", currentTx.Version, currentTx.ID)
+		case <-requestLatestTxn:
```

Comment:
I thought of those two solutions. 

1. Loop over the channel and retrieve latest event while necessary. 

    A lot of events could be filled up in the states pub-sub, because off-chain states can be generated in large numbers, while the registered event will happen only when closing the channel or in case of dispute.

2. Use a mutex

    In this case, we will still require a go-routine that keeps draining the pub-sub. In go-routine, every time a new state is received, the mutex will have to be locked and currentTx should be updated.
    
    While the latest state is necessary, the mutex will have to be locked and currentTx should be read.
   
    So, this case would look more or less like the "receive latest event" logic, with an additional overhead that the mutex lock should be acquired each time a new state is received.

In go, it is idiomatic to share values by communicating over channels, rather than by using shared memory + mutex (see [here](https://golang.org/doc/effective_go#sharing)). The advantage is that, one need not worry about locking and unlocking the mutex in a proper way (that avoids dead-locks and race conditions).

Reply:
Also, i have refactored this part and extracted the two (request, response) channels to a separate type. See 576dce3a8c296a672a0f851a38d155d621f197ff